lib/ptracer.h

```diff
@@ -0,0 +1,471 @@
+/* 
+ * Copyright (c) 2024, The Pallene Developers
+ * Pallene Tracer is licensed under the BSD-3-Clause license.
+ * Please refer to the LICENSE and AUTHORS files for details
+ * SPDX-License-Identifier: BSD-3-Clause 
+ */
+
+#ifndef PALLENE_TRACER_H
+#define PALLENE_TRACER_H
+
+#include <lua.h>
+#include <lauxlib.h>
+#include <luacore.h>
+
+#include <stdlib.h>
+#include <stdbool.h>
+#include <string.h>
+
+/* ---------------- MACRO DEFINITIONS ---------------- */
+
+#if defined(__GNUC__) || defined(__clang__) 
+#define PALLENE_TRACER_UNREACHABLE    __builtin_unreachable()
+#elif defined(_MSC_VER) // MSVC
+#define PALLENE_TRACER_UNREACHABLE    __assume(false)
+#endif
+
+/* Pallene stack reference entry for the registry. */
+/* DO NOT CHANGE EVEN BY MISTAKE. */
+#define PALLENE_TRACER_CONTAINER_ENTRY  "__PALLENE_TRACER_CONTAINER"
+
+/* Finalizer metatable key. */
+/* DO NOT CHANGE EVEN BY MISTAKE. */
+#define PALLENE_TRACER_FINALIZER_ENTRY  "__PALLENE_TRACER_FINALIZER"
+
+/* The size of the Pallene call-stack. */
+/* DO NOT CHANGE EVEN BY MISTAKE. */
+#define PALLENE_TRACER_MAX_CALLSTACK           2000000
+
+/* Traceback ellipsis top threshold. How many frames should we print
+   first to trigger ellipsis? */
+#define PALLENE_TRACEBACK_TOP_THRESHOLD      10
+/* This should always be 2 fewer than top threshold, for symmetry.
+   Becuase we will always have 2 tail frames lingering around at
+   at the end which is not captured by '_countlevels'. Lua also 
+   do it like this. */
+#define PALLENE_TRACEBACK_BOTTOM_THRESHOLD    8
+
+/* ---------------- MACRO DEFINITIONS END ---------------- */
+
+/* ---------------- DATA STRUCTURES ---------------- */
+
+/* What type of frame we are dealing with? Is it just a normal
+   C function or Lua C Function? */
+typedef enum frame_type {
+    PALLENE_TRACER_FRAME_TYPE_C,
+    PALLENE_TRACER_FRAME_TYPE_LUA
+} frame_type_t;
+
+/* Details of the callee function (name, where it is from etc.) */
+/* Optimization Tip: Try making it 'static' for every funciton. */
+typedef struct pt_fn_details {
+    const char *const fn_name;
+    const char *const filename;
+} pt_fn_details_t;
+
+/* A single frame representation. */
+typedef struct pt_frame {
+    frame_type_t type;
+    int line;
+
+    union {
+        pt_fn_details_t *details;
+        lua_CFunction frame_sig;
+    } shared;
+} pt_frame_t;
+
+/* Out stack is fully heap-allocated stack. We need some structure to hold
+   the stack information. This structure will be an Userdatum. */
+/* 'cont' stands for 'container'. */
+typedef struct pt_cont {
```

Comment:
Just call it pt_containiner then :)
My first thought for cont is continuation.

Reply:
The comment exist to clarify. Simple is better.